<!-- Launches all necessary vision applications. -->

<launch>
    <param name="cropx" value="95" />
    <param name="cropy" value="145" />
    <param name="cropwidth" value="440" />
    <param name="cropheight" value="400"/>
    <!-- Launch stereo and mono converters, depending on which method you want to use for gathering points -->
    <node pkg="stereo_click" type="mono_converter.py" name="mono_converter_node" output="screen">
		<param name="input" value="processing_points" />
		<param name="output" value="stereo_points_3d" />
		<param name="output_frame" value="base_footprint" />
	</node>
	
	<node pkg="stereo_click" type="stereo_converter.py" name="stereo_converter_node" output="screen">
		<param name="input1" value="left_processing_points" />
		<param name="input2" value="right_processing_points" />
		<param name="output" value="stereo_points_3d" />
		<param name="output_frame" value="base_footprint" />
	</node>

    <!-- Image Processor nodes for finding the furthest corners, center of towel clump, and fitting triangle and towel models -->
    <node pkg="image_processor" type="triangle_fitter_node.py" name="triangle_fitter_node" output="screen">
      <env name="PYTHONPATH" value="$(find clothing_models)/src/clothing_models:$(find visual_feedback_utils)/../thirdparty/python-cvgreyc/cvgreyc/build/lib.linux-x86_64-2.6:$(optenv PYTHONPATH)"/>
        <param name="mono_converter" value="mono_converter_node" />
        <param name="stereo_converter" value="stereo_converter_node" />
        <param name="threshold" value="100" />
        <param name="model" value="triangles" />
        <param name="num_iters" value="15" />
    </node>
    
    <node pkg="image_processor" type="shape_fitter_node.py" name="towel_fitter_node" output="screen" >
      <env name="PYTHONPATH" value="$(find clothing_models)/src/clothing_models:$(find visual_feedback_utils)/../thirdparty/python-cvgreyc/cvgreyc/build/lib.linux-x86_64-2.6:$(optenv PYTHONPATH)"/>
        <param name="mono_converter" value="mono_converter_node" />
        <param name="stereo_converter" value="stereo_converter_node" />
        <param name="threshold" value="100" />
        <param name="model" value="towel_model" />
        <param name="mode" value="towel" />
        <param name="num_iters" value="15" />
    </node>
    
    <node pkg="image_processor" type="furthest_corner_node.py" name="far_right_finder_node" output="screen" >
        <param name="mono_converter" value="mono_converter_node" />
        <param name="stereo_converter" value="stereo_converter_node" />
        <param name="threshold" value="100" />
        <param name="left_to_right" value="true" />
    </node>
    
    <node pkg="image_processor" type="furthest_corner_node.py" name="far_left_finder_node" output="screen" >
        <param name="mono_converter" value="mono_converter_node" />
        <param name="stereo_converter" value="stereo_converter_node" />
        <param name="threshold" value="100" />
        <param name="left_to_right" value="false" />
    </node>
    <node pkg="image_processor" type="clump_center_node.py" name="clump_center_node" output="screen" >
        <param name="mono_converter" value="mono_converter_node" />
        <param name="stereo_converter" value="stereo_converter_node" />
        <param name="threshold" value="100" />
    </node>
    
    <!-- View the annotated feed for said processors -->
    <node pkg="image_view" type="image_view" name="clump_center_viewer">
        <remap from="image" to="clump_center_node/annotated" />
    </node>
    <node pkg="image_view" type="image_view" name="left_corner_viewer">
        <remap from="image" to="far_left_finder_node/annotated" />
    </node>
    
    <node pkg="image_view" type="image_view" name="right_corner_viewer">
        <remap from="image" to="far_right_finder_node/annotated" />
    </node>
    
    <node pkg="image_view" type="image_view" name="triangle_viewer">
        <remap from="image" to="triangle_fitter_node/annotated" />
    </node>
    
    <node pkg="image_view" type="image_view" name="towel_viewer">
        <remap from="image" to="towel_fitter_node/annotated" />
    </node>

</launch>
